src/sagemaker_xgboost_container/metrics/custom_metrics.py

```diff
@@ -52,6 +52,39 @@ def accuracy(preds, dtrain):
 
 def f1(preds, dtrain):
     """Compute f1 score. This can be used for multiclassification training.
+    For more information see: https://scikit-learn.org/stable/modules/generated/sklearn.metrics.f1_score.html
+    :param preds: Prediction values
+    :param dtrain: Training data with labels
+    :return: Metric name, f1 score
+    """
+    score = 0.0
+    if preds.size > 0:
+        labels = dtrain.get_label()
+        pred_labels = margin_to_class_label(preds)
+        score = f1_score(labels, pred_labels, average='macro')
+    return 'f1', score
+
+
+def f1_binary(preds, dtrain):
+    """Compute f1 binary score. This can be used for binaryclassification training.
+
+    For more information see: https://scikit-learn.org/stable/modules/generated/sklearn.metrics.f1_score.html
+
+    :param preds: Prediction values
+    :param dtrain: Training data with labels
+    :return: Metric name, f1 score
+    """
+    score = 0.0
+    if preds.size > 0:
+        labels = dtrain.get_label()
+        pred_labels = margin_to_class_label(preds)
+        # this function is used only for AutoPilot and the least frequent label is already encoded as 1.
```

Comment:
There's nothing that stops external customers from using this directly, so I wouldn't make this assumption. 
Also I'm not sure what the implications of the 'least frequent label is already encoded as 1' is. Can you restate the assumption here to ensure behavior is clear. 

Also do we need validation for the number of labels to be `2`?

Reply:
Why does this assumption matter? For F1, we just care what's positive/negative.

Reply:
To make f1 binary work, the only requirement is the label is encoded in {0, 1}. Taking least frequent label as 1 is only a behavior AutoPilot enforces. The customer could also use more frequent label as 1.

A question here is that are the labels already encoded in {0, 1}? If not, we need to have another label encoder to make f1 binary works.

Reply:
@huibinshen Note that you get the labels from `margin_to_class_label` function, which always encodes in `{0,1}`. So your code is fine. More importantly, XGBoost is a product itself. You shouldn't add comments about Autopilot's behavior here. Just remove the comment, which is the root cause of confusions.

Reply:
Thanks Haifeng, will remove the comment in the next revision.

@iyerr3 Do you have other comments? I hope the explanation above is clear.

Reply:
Thanks for removing the comment. 
My question on validation still stands: if customer uses `f1_binary` with multi-class, then they should get a Customer Error. Is that validation happening somewhere that I missed?

Reply:
Hi Rahul, I added a customer error in the new revision.